winpr/libwinpr/clipboard/posix.c

```diff
@@ -572,28 +584,118 @@ static FILEDESCRIPTORW* convert_local_file_list_to_filedescriptors(wArrayList* f
 	return NULL;
 }
 
+static void* convert_any_uri_list_to_filedescriptors(wClipboard* clipboard, UINT32 formatId,
+                                                     UINT32* pSize)
+{
+	FILEDESCRIPTORW* descriptors =
+	    convert_local_file_list_to_filedescriptors(clipboard->localFiles);
+
+	WINPR_ASSERT(pSize);
+
+	*pSize = 0;
+	if (!descriptors)
+		return NULL;
+
+	*pSize = (UINT32)ArrayList_Count(clipboard->localFiles) * sizeof(FILEDESCRIPTORW);
+	clipboard->fileListSequenceNumber = clipboard->sequenceNumber;
+	return descriptors;
+}
+
 static void* convert_uri_list_to_filedescriptors(wClipboard* clipboard, UINT32 formatId,
                                                  const void* data, UINT32* pSize)
 {
-	FILEDESCRIPTORW* descriptors = NULL;
+	const UINT32 expected = ClipboardGetFormatId(clipboard, mime_uri_list);
+	if (formatId != expected)
+		return NULL;
+	if (!process_uri_list(clipboard, (const char*)data, *pSize))
+		return NULL;
+	return convert_any_uri_list_to_filedescriptors(clipboard, formatId, pSize);
+}
 
-	if (!clipboard || !data || !pSize)
+static BOOL process_files(wClipboard* clipboard, const char* data, UINT32 pSize, const char* prefix)
+{
+	const size_t prefix_len = strlen(prefix);
+
+	WINPR_ASSERT(clipboard);
+	if (!data || (pSize < prefix_len))
+		return FALSE;
+	if (strncmp(data, prefix, prefix_len) != 0)
+		return FALSE;
+	data += prefix_len;
+	pSize -= prefix_len;
+
+	BOOL rc = FALSE;
+	char* copy = strndup(data, pSize);
+	if (!copy)
+		goto fail;
+
+	char* endptr = NULL;
+	char* tok = strtok_s(copy, "\n", &endptr);
+	while (tok)
+	{
+		size_t tok_len = strnlen(tok, pSize);
+		if (!process_uri(clipboard, tok, tok_len))
+			goto fail;
+		pSize -= tok_len;
+		tok = strtok_s(NULL, "\n", &endptr);
+	}
+	rc = TRUE;
+
+fail:
+	free(copy);
+	return rc;
+}
+
+static BOOL process_gnome_copied_files(wClipboard* clipboard, const char* data, UINT32 pSize)
+{
+	return process_files(clipboard, data, pSize, "copy\n");
+}
+
+static BOOL process_mate_copied_files(wClipboard* clipboard, const char* data, UINT32 pSize)
+{
+	return process_files(clipboard, data, pSize, "copy\n");
+}
+
+static BOOL process_nautilus_clipboard(wClipboard* clipboard, const char* data, UINT32 pSize)
+{
+	return process_files(clipboard, data, pSize, "x-special/nautilus-clipboard\copy\n");
```

Comment:
I wonder, when we will remove the old nautilus-clipboard type. Next week GNOME 43 is released (meaning only 42 and 43 are supported as stable releases). Then old nautilus-clipboard type was removed in 40.
Maybe we can remove it here in FreeRDP with the next Debian release next year (which will use GNOME 43 for debian stable).

---

In any case, `"x-special/nautilus-clipboard\copy\n"` should probably be `"x-special/nautilus-clipboard\\copy\n"` (missing escaping).